NullPointerException

Clicking on a framebuffer in the filmstrip produced this exception:

```
[main][Main$UI$1.handleException] Unhandled exception in the UI thread.
[main][Main$UI$1.handleException] org.eclipse.swt.SWTException: Failed to execute runnable (java.lang.NullPointerException)
[main][Main$UI$1.handleException]    at org.eclipse.swt.SWT.error(Unknown Source)
[main][Main$UI$1.handleException]    at org.eclipse.swt.SWT.error(Unknown Source)
[main][Main$UI$1.handleException]    at org.eclipse.swt.widgets.Synchronizer.runAsyncMessages(Unknown Source)
[main][Main$UI$1.handleException]    at org.eclipse.swt.widgets.Display.runAsyncMessages(Unknown Source)
[main][Main$UI$1.handleException]    at org.eclipse.swt.widgets.Display.readAndDispatch(Unknown Source)
[main][Main$UI$1.handleException]    at org.eclipse.jface.window.Window.runEventLoop(Window.java:818)
[main][Main$UI$1.handleException]    at org.eclipse.jface.window.Window.open(Window.java:794)
[main][Main$UI$1.handleException]    at com.google.gapid.Main$UI.show(Main.java:107)
[main][Main$UI$1.handleException]    at com.google.gapid.Main.main(Main.java:68)
[main][Main$UI$1.handleException] Caused by: java.lang.NullPointerException
[main][Main$UI$1.handleException]    at org.eclipse.jface.viewers.CustomHashtable.hashCode(CustomHashtable.java:268)
[main][Main$UI$1.handleException]    at org.eclipse.jface.viewers.CustomHashtable.get(CustomHashtable.java:240)
[main][Main$UI$1.handleException]    at org.eclipse.jface.viewers.TreeSelection$InitializeData.<init>(TreeSelection.java:61)
[main][Main$UI$1.handleException]    at org.eclipse.jface.viewers.TreeSelection$InitializeData.<init>(TreeSelection.java:53)
[main][Main$UI$1.handleException]    at org.eclipse.jface.viewers.TreeSelection.<init>(TreeSelection.java:98)
[main][Main$UI$1.handleException]    at org.eclipse.jface.viewers.TreeSelection.<init>(TreeSelection.java:109)
[main][Main$UI$1.handleException]    at com.google.gapid.views.AtomTree.lambda$onAtomsSelected$7(AtomTree.java:428)
[main][Main$UI$1.handleException]    at com.google.gapid.util.SelectionHandler$1.lambda$onSuccess$0(SelectionHandler.java:75)
[main][Main$UI$1.handleException]    at com.google.gapid.widgets.Widgets.ifNotDisposed(Widgets.java:113)
[main][Main$UI$1.handleException]    at com.google.gapid.widgets.Widgets.lambda$scheduleIfNotDisposed$0(Widgets.java:125)
[main][Main$UI$1.handleException]    at com.google.gapid.widgets.Widgets.lambda$scheduleIfNotDisposed$1(Widgets.java:134)
[main][Main$UI$1.handleException]    at org.eclipse.swt.widgets.RunnableLock.run(Unknown Source)
[main][Main$UI$1.handleException]    ... 7 more
```

Repro steps:
* Open capture attached in https://github.com/google/gapid/issues/479.
* Select the last command tree node `278010: OpenGL ES context 3`
* Click on frame `546`

--- a/gapic/src/main/com/google/gapid/models/AtomStream.java
+++ b/gapic/src/main/com/google/gapid/models/AtomStream.java
@@ -20,6 +20,7 @@ import static com.google.gapid.util.Paths.any;
 import static com.google.gapid.util.Paths.commandTree;
 import static com.google.gapid.util.Paths.lastCommand;
 import static com.google.gapid.util.Paths.observationsAfter;
+import static java.util.logging.Level.FINE;
 
 import com.google.common.base.Objects;
 import com.google.common.util.concurrent.Futures;
@@ -52,7 +53,7 @@ import java.util.logging.Logger;
  */
 public class AtomStream extends ModelBase.ForPath<AtomStream.Node, Void, AtomStream.Listener>
     implements ApiContext.Listener, Capture.Listener {
-  private static final Logger LOG = Logger.getLogger(AtomStream.class.getName());
+  protected static final Logger LOG = Logger.getLogger(AtomStream.class.getName());
 
   private final Capture capture;
   private final ApiContext context;
@@ -240,7 +241,9 @@ public class AtomStream extends ModelBase.ForPath<AtomStream.Node, Void, AtomStr
       @Override
       protected Path.CommandTreeNode onRpcThread(Result<Value> result)
           throws RpcException, ExecutionException {
-        return result.get().getPath().getCommandTreeNode();
+        Service.Value value = result.get();
+        LOG.log(FINE, "Resolved selection to {0}", value);
+        return value.getPath().getCommandTreeNode();
       }
 
       @Override
@@ -450,6 +453,10 @@ public class AtomStream extends ModelBase.ForPath<AtomStream.Node, Void, AtomStr
         node = children[child] = new Node(this, child);
       }
       return node;
+    }
+
+    public boolean isLastChild() {
+      return parent == null || (parent.getChildCount() - 1 == index);
     }
 
     public CommandTreeNode getData() {

--- a/gapic/src/main/com/google/gapid/views/AtomTree.java
+++ b/gapic/src/main/com/google/gapid/views/AtomTree.java
@@ -376,7 +376,7 @@ public class AtomTree extends Composite implements Tab, Capture.Listener, AtomSt
 
   private void processSearchResult(Service.FindResponse found) {
     ListenableFuture<TreePath> path = getTreePath(models.atoms.getData(), Lists.newArrayList(),
-        found.getCommandTreeNode().getIndicesList().iterator(), false);
+        found.getCommandTreeNode().getIndicesList().iterator());
     Rpc.listen(path, searchController, new UiCallback<TreePath, TreePath>(viewer.getTree(), LOG) {
       @Override
       protected TreePath onRpcThread(Result<TreePath> result)
@@ -466,12 +466,27 @@ public class AtomTree extends Composite implements Tab, Capture.Listener, AtomSt
   }
 
   private ListenableFuture<TreePath> getTreePath(AtomIndex index) {
-    return getTreePath(models.atoms.getData(), Lists.newArrayList(),
-        index.getNode().getIndicesList().iterator(), index.isGroup());
+    AtomStream.Node root = models.atoms.getData();
+    ListenableFuture<TreePath> result = getTreePath(root, Lists.newArrayList(root),
+        index.getNode().getIndicesList().iterator());
+    if (index.isGroup()) {
+      // Find the deepest group/node in the path that is not the last child of its parent.
+      result = Futures.transform(result, path -> {
+        while (path.getSegmentCount() > 0) {
+          AtomStream.Node node = (AtomStream.Node)path.getLastSegment();
+          if (!node.isLastChild()) {
+            break;
+          }
+          path = path.getParentPath();
+        }
+        return path;
+      });
+    }
+    return result;
   }
 
   private ListenableFuture<TreePath> getTreePath(
-      AtomStream.Node node, List<Object> path, Iterator<Long> indices, boolean group) {
+      AtomStream.Node node, List<Object> path, Iterator<Long> indices) {
     ListenableFuture<AtomStream.Node> load = models.atoms.load(node);
     if (!indices.hasNext()) {
       TreePath result = new TreePath(path.toArray());
@@ -479,21 +494,17 @@ public class AtomTree extends Composite implements Tab, Capture.Listener, AtomSt
       return (load == null) ? Futures.immediateFuture(result) :
           Futures.transform(load, ignored -> result);
     }
-    return (load == null) ? getTreePathForLoadedNode(node, path, indices, group) :
-        Futures.transformAsync(
-            load, loaded -> getTreePathForLoadedNode(loaded, path, indices, group));
+    return (load == null) ? getTreePathForLoadedNode(node, path, indices) :
+        Futures.transformAsync(load, loaded -> getTreePathForLoadedNode(loaded, path, indices));
   }
 
   private ListenableFuture<TreePath> getTreePathForLoadedNode(
-      AtomStream.Node node, List<Object> path, Iterator<Long> indices, boolean group) {
+      AtomStream.Node node, List<Object> path, Iterator<Long> indices) {
     int index = indices.next().intValue();
-    if (group && index == node.getChildCount() - 1) {
-      return Futures.immediateFuture(new TreePath(path.toArray()));
-    }
 
     AtomStream.Node child = node.getChild(index);
     path.add(child);
-    return getTreePath(child, path, indices, group);
+    return getTreePath(child, path, indices);
   }
 
   /**

